File: go-service/internal/application/usecase/update_product.go
package usecase

import (
	"context"

	appcache "github.com/example/interview-stack/go-service/internal/application/cache"
	"github.com/example/interview-stack/go-service/internal/application/dto"
	"github.com/example/interview-stack/go-service/internal/domain"
)

// UpdateProductUseCase updates an existing product.
type UpdateProductUseCase struct {
	repo         domain.ProductRepository
	logger       domain.Logger
	clock        domain.Clock
	queue        domain.WriteQueue
	writeBehind  bool
	source       string
	cacheWriter  cacheWriter
	cacheEnabled bool
}

func NewUpdateProductUseCase(
	repo domain.ProductRepository,
	logger domain.Logger,
	clock domain.Clock,
	cacheSvc *appcache.Service,
	queue domain.WriteQueue,
	writeBehind bool,
	source string,
) *UpdateProductUseCase {
	return &UpdateProductUseCase{
		repo:         repo,
		logger:       logger,
		clock:        clock,
		queue:        queue,
		writeBehind:  writeBehind,
		source:       source,
		cacheWriter:  newCacheWriter(cacheSvc, logger),
		cacheEnabled: cacheSvc != nil && cacheSvc.Enabled(),
	}
}

func (uc *UpdateProductUseCase) Execute(ctx context.Context, id string, input dto.UpdateProductRequest) (*dto.ProductResponse, error) {
	product := input.ToDomain()
	if err := product.Validate(); err != nil {
		return nil, err
	}
	if uc.writeBehind && uc.queue != nil {
		current, err := uc.repo.GetProduct(ctx, id)
		if err != nil {
			return nil, err
		}
		if current == nil {
			return nil, domain.NewNotFoundError("product not found")
		}
		current.Name = product.Name
		current.Price = product.Price
		current.Tags = product.Tags
		current.UpdatedAt = uc.clock.Now()
		dtoProduct := dto.FromDomainProduct(current)
		uc.cacheWriter.upsertProduct(ctx, &dtoProduct)
		uc.cacheWriter.invalidateList(ctx)
		payload := *current
		event := domain.WriteEvent{
			Type:      domain.WriteEventUpdate,
			ID:        id,
			Payload:   &payload,
			Timestamp: current.UpdatedAt,
			Source:    uc.source,
		}
		if err := uc.queue.Enqueue(ctx, event); err != nil {
			return nil, domain.NewInternalError("failed to enqueue update event", err)
		}
		uc.logger.Info("product enqueued for update", domain.KV("id", id))
		return &dtoProduct, nil
	}

	product.UpdatedAt = uc.clock.Now()

	updated, err := uc.repo.UpdateProduct(ctx, id, product)
	if err != nil {
		return nil, err
	}
	if updated == nil {
		return nil, domain.NewNotFoundError("product not found")
	}
	dtoProduct := dto.FromDomainProduct(updated)
	if uc.cacheEnabled {
		uc.cacheWriter.upsertProduct(ctx, &dtoProduct)
		uc.cacheWriter.invalidateList(ctx)
	}
	uc.logger.Info("product updated", domain.KV("id", id))
	return &dtoProduct, nil
}

File: node-service/internal/application/usecases/updateProduct.js
const DomainError = require('../../domain/errors');
const { toDomain, toResponse } = require('../dto/productDto');

class UpdateProductUseCase {
  /**
   * @param {import('../../domain/productRepository')} repository
   * @param {import('../../domain/logger')} logger
   * @param {import('../../core/clock/Clock')} clock
   * @param {import('../services/cacheCoordinator')} cacheCoordinator
   * @param {import('../services/writeQueue')} writeQueue
   */
  constructor(repository, logger, clock, cacheCoordinator, writeQueue, writeBehindEnabled, source) {
    this.repository = repository;
    this.logger = logger;
    this.clock = clock;
    this.cache = cacheCoordinator;
    this.queue = writeQueue;
    this.writeBehind = Boolean(writeBehindEnabled);
    this.source = source || 'node-service';
  }

  async execute(id, input) {
    const changes = toDomain({ ...input, id });
    changes.validate();
    const now = this.clock.now();
    changes.updatedAt = now;

    if (this.writeBehind && this.queue) {
      const current = await this.repository.getById(id);
      if (!current) {
        throw DomainError.notFound('product not found');
      }
      current.name = changes.name;
      current.price = changes.price;
      current.tags = changes.tags;
      current.updatedAt = now;
      const response = toResponse(current);
      await this.cache?.upsertProduct(response);
      await this.enqueueEvent('update', current, now);
      this.logger.info('product enqueued for update', { id });
      return response;
    }

    const updated = await this.repository.update(id, changes);
    if (!updated) {
      throw DomainError.notFound('product not found');
    }
    const response = toResponse(updated);
    await this.cache?.upsertProduct(response);
    this.logger.info('product updated', { id });
    return response;
  }

  async enqueueEvent(type, product, timestamp) {
    const event = {
      type,
      id: product.id,
      payload: product,
      timestamp: timestamp?.toISOString?.() || new Date().toISOString(),
      source: this.source
    };
    await this.queue.enqueue(event);
  }
}

module.exports = UpdateProductUseCase;
